Disable beatmap skinning support when entering the skin editor

peppy opened the pull request:
As proposed in https://github.com/ppy/osu/issues/23915.

Closes #23915 
Closes #22561

This *definitely* needs test coverage, but wanted to get this out to make sure people agree on the behaviour and direction before considering adding them.

peppy committed:
Disable beatmap skinning when entering the skin editor

bdach commented on osu.Game/Overlays/SkinEditor/SkinEditorOverlay.cs:
```diff
@@ -147,17 +156,24 @@ public void OnReleased(KeyBindingReleaseEvent<GlobalAction> e)
 
-            // AddOnce with parameter will ensure the newest target is loaded if there is any overlap.
-            Scheduler.AddOnce(setTarget, screen);
+                // AddOnce with parameter will ensure the newest target is loaded if there is any overlap.
+                Scheduler.AddOnce(setTarget, screen);
+            }
+            finally
+            {
+                globallyReenableBeatmapSkinSetting();
+            }
```
So `setTarget()` - which internally contains a `globallyDisableBeatmapSkinSetting()` call if the target screen is a `Player` - is scheduled, but then `globallyReenableBeatmapSkinSetting()` is called in a `finally` block? It's probably not *wrong*, as the schedule saves this, but why is this written like that in the first place? I assume the `finally` is aimed at the early-return, but if so, then can the re-enable not be done at the start of the method or something?

bdach commented on osu.Game/Overlays/SkinEditor/SkinEditorOverlay.cs:
```diff
@@ -182,5 +201,30 @@ private void setTarget(OsuScreen? target)
+            if (beatmapSkins.Disabled)
+                return;
+
+            // The skin editor doesn't work well if beatmap skins are being applied to the player screen.
+            // To keep things simple, disable the setting game-wide while using the skin editor.
+            beatmapSkinsOriginalState = beatmapSkins.Value;
+
+            beatmapSkins.Value = false;
+            beatmapSkins.Disabled = true;
```
Is there a reason why this is essentially manually reimplementing what a bindable lease is?

```diff
diff --git a/osu.Game/Overlays/SkinEditor/SkinEditorOverlay.cs b/osu.Game/Overlays/SkinEditor/SkinEditorOverlay.cs
index b120faa45f..4f0028de64 100644
--- a/osu.Game/Overlays/SkinEditor/SkinEditorOverlay.cs
+++ b/osu.Game/Overlays/SkinEditor/SkinEditorOverlay.cs
@@ -203,7 +203,7 @@ private void setTarget(OsuScreen? target)
         }
 
         private readonly Bindable<bool> beatmapSkins = new Bindable<bool>();
-        private bool beatmapSkinsOriginalState;
+        private LeasedBindable<bool>? leasedBeatmapSkins;
 
         private void globallyDisableBeatmapSkinSetting()
         {
@@ -212,19 +212,14 @@ private void globallyDisableBeatmapSkinSetting()
 
             // The skin editor doesn't work well if beatmap skins are being applied to the player screen.
             // To keep things simple, disable the setting game-wide while using the skin editor.
-            beatmapSkinsOriginalState = beatmapSkins.Value;
-
-            beatmapSkins.Value = false;
-            beatmapSkins.Disabled = true;
+            leasedBeatmapSkins = beatmapSkins.BeginLease(true);
+            leasedBeatmapSkins.Value = false;
         }
 
         private void globallyReenableBeatmapSkinSetting()
         {
-            if (!beatmapSkins.Disabled)
-                return;
-
-            beatmapSkins.Disabled = false;
-            beatmapSkins.Value = beatmapSkinsOriginalState;
+            leasedBeatmapSkins?.Return();
+            leasedBeatmapSkins = null;
         }
     }
 }

```

bdach commented:
Additionally, if the skin editor is opened and exited on a beatmap with beatmap skins active, the skin editor will turn off beatmap skins and disable the toggle on open, but it will not re-enable and restore the previous value on exit:

https://github.com/ppy/osu/assets/20418176/9c40ea53-de42-462b-98ce-5b6a76d47d68

which I would say is bad.

peppy commented:
> Additionally, if the skin editor is opened and exited on a beatmap with beatmap skins active, the skin editor will turn off beatmap skins and disable the toggle on open, but it will not re-enable and restore the previous value on exit:
> 
> 
> 
> https://github.com/ppy/osu/assets/20418176/9c40ea53-de42-462b-98ce-5b6a76d47d68
> 
> 
> 
> which I would say is bad.

Are you expecting it to re-enable without exiting player? Because that wasn't my intention with implementation (would require the skin editor to save changes in toggle rather than screen change).

bdach commented:
> Are you expecting it to re-enable without exiting player? Because that wasn't my intention with implementation (would require the skin editor to save changes in toggle rather than screen change).

I suppose so. It feels kinda janky that this entirely unrelated action of opening the skin editor causes beatmap skins to be disabled (for good reasons) but then doesn't put things back as they were once you're done editing your personal skin. Not sure.

peppy commented:
I guess we can go in that direction. What do you think about providing confirmation to save when exiting the editor in that case? Should it be something we expedite since it's happening a lot more often?

bdach commented:
> What do you think about providing confirmation to save when exiting the editor in that case?

I kinda think that's not really needed. It can just silently save the same way the skin editor already does on screen change.

peppy commented:
Okay, I'll give that a try. It may be okay now that saving skins is a lot faster (in the past there was a significant overhead).

peppy committed:
Use bindable lease instead of reimplementing the same thing locally

peppy committed:
Change skin editor flow to always save on toggle

This also moves the beatmap skin disable toggle to on toggle, in line
with review feedback.

I've decided to always apply the disable, not just on the `Player`
screen. It should be assumed that if a user is in the skin editor they
are never going to need access to this anyway.

peppy commented:
I've made the requested changes. Feels pretty good IMO.

bdach committed:
Remove unused using directive

bdach committed:
Merge branch 'master' into skin-editor-beatmap-skin-disable
